### ISSUE TITLE
Isn't possible validate if "0" (zero as string) is a string with "notEmpty" method.

### ISSUE THREAD

**Issue by bsmrs** [Respect/Validation#438]
This test will return an exception:

```php
Validator::type('string')
    ->notEmpty()
    ->check("0");
```

The tests in "library/Rules/NotEmpty.php" don't cover this case.

**Comment by henriquemoody** [comment]
On version 1.0, which will be released on Oct 24th, all rules will be mandatory, so if your want to handle optional values (`null` or `''`), you will use [optional()](https://github.com/Respect/Validation/blob/master/docs/Optional.md) rule.

To be honest I intent to refactor this rule and create a NotBlank rule. They will behave like this:

Value           | NotEmpty | NotBlank
----------------|----------|------
 `null`         | `FALSE`  | `FALSE`
 `""`           | `FALSE`  | `FALSE`
 `[]`           | `FALSE`  | `FALSE`
 `" "`          | `TRUE`   | `FALSE`
 `0`            | `FALSE`  | `FALSE`
 `"0"`          | `FALSE`  | `FALSE`
 `0.0`          | `FALSE`  | `FALSE`
 `"0.0"`        | `TRUE`   | `FALSE`
 `false`        | `FALSE`  | `FALSE`
 `[""]`         | `TRUE`   | `FALSE`
 `[" "]`        | `TRUE`   | `FALSE`
 `[0]`          | `TRUE`   | `FALSE`
 `["0"]`        | `TRUE`   | `FALSE`
 `[false]`      | `TRUE`   | `FALSE`
 `[[""], [0]]`  | `TRUE`   | `FALSE`
 `-1`           | `TRUE`   | `TRUE`


What you think?

**Comment by bsmrs** [comment]
I thought about "notBlank" method and I liked it.

I think just the notEmpty method must be able to distinct between empty string ("") and  "0". Validating an integer with method “Validator::type('int')" already throws an exception for "" (empty string) because it isn't an integer. So we don't need to check it again in methods "notEmpty" or "notBlank”.

Thanks for the feedback.

**Comment by henriquemoody** [comment]
So, if you think notEmpty() should distinct between an empty string and "0" (as an string), what values do you think should be considered _empty_?

**Comment by bsmrs** [comment]
I think the same values that you put in table on your first answer, changing just "0". The remaining question is the 0 (zero as int), this is a valid integer and notEmpty method will return false, but we can validate it using “Validator::type('int')". Otherwise, the method notEmpty could use "is_int" function to check integers and return true for 0 (0 as int).

Both are right in my opinion, adjust just for string or for integer too.

**Comment by henriquemoody** [comment]
I see... As I said on #439, I have plans to give NotEmpty rule the same opposite behavior of PHP's empty() language construct because it sounds the reasonable thing to do to me. You said zero must not be considered empty, but how about `false` or 0.0 as float, or an string with zeros? It's complicated and that's the reason why I created [NotBlank](https://github.com/Respect/Validation/blob/880cdb/docs/NotBlank.md) rule, it's very restrictive, but it's very useful, IMO.

**Comment by henriquemoody** [comment]
BTW, the code on the description of this issue can not simply be done on version 1.0 with:

```php
Validator::stringType()->check("0");
```

No exception will be thrown.

**Comment by henriquemoody** [comment]
- **NotBlank**: empty string, string with white spaces only, `false`, `null`, zeros and arrays the only these values
- **NotEmpty**: the opposite behavior of PHP's empty()
- **NotOptional**: the optional values for input, which I consider as an empty string (`''`) and null (`null`), only.

I would like to head about you guys what you think about it:

Value           | NotOptional | NotEmpty | NotBlank
----------------|-------------|----------|------
 `null`         | `FALSE`     | `FALSE`  | `FALSE`
 `""`           | `FALSE`     | `FALSE`  | `FALSE`
 `[]`           | `TRUE`      | `FALSE`  | `FALSE`
 `" "`          | `TRUE`      | `TRUE`   | `FALSE`
 `0`            | `TRUE`      | `FALSE`  | `FALSE`
 `"0"`          | `TRUE`      | `FALSE`  | `FALSE`
 `0.0`          | `TRUE`      | `FALSE`  | `FALSE`
 `"0.0"`        | `TRUE`      | `TRUE`   | `FALSE`
 `false`        | `TRUE`      | `FALSE`  | `FALSE`
 `[""]`         | `TRUE`      | `TRUE`   | `FALSE`
 `[" "]`        | `TRUE`      | `TRUE`   | `FALSE`
 `[0]`          | `TRUE`      | `TRUE`   | `FALSE`
 `["0"]`        | `TRUE`      | `TRUE`   | `FALSE`
 `[false]`      | `TRUE`      | `TRUE`   | `FALSE`
 `[[""], [0]]`  | `TRUE`      | `TRUE`   | `FALSE`
 `-1`           | `TRUE`      | `TRUE`   | `TRUE`

What you think, @bsmrs?

**Comment by bsmrs** [comment]
It seems me very good.  With those three methods we can validate everything.

**Comment by henriquemoody** [comment]
[NotOptional](https://github.com/Respect/Validation/blob/15b0a9/docs/NotOptional.md) was created and will be available on version 1.0 which will be released on Oct 24th.
